File: encodify/Classes/Utilities/Controller/ImageEncodeViewController.m
//
//  ImageEncodeViewController.m
//  encodify
//
//  Created by 洪朔 on 2016/11/30.
//  Copyright © 2016年 Tuccuay. All rights reserved.
//

#import "ImageEncodeViewController.h"

#import "EToast.h"

#import <Masonry/Masonry.h>

@interface ImageEncodeViewController ()<UINavigationControllerDelegate, UIImagePickerControllerDelegate>

@property (strong, nonatomic) UITextView *textView;

@end

@implementation ImageEncodeViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.view.backgroundColor = [UIColor whiteColor];
    
    UIImagePickerController *imagePickerController = [[UIImagePickerController alloc] init];
    imagePickerController.sourceType = UIImagePickerControllerSourceTypePhotoLibrary;
    imagePickerController.delegate = self;
    [self presentViewController:imagePickerController animated:YES completion:nil];
    
    self.navigationItem.rightBarButtonItem = [[UIBarButtonItem alloc] initWithTitle:@"Copy" style:UIBarButtonItemStylePlain target:self action:@selector(copyResult)];
}

- (void)copyResult {
    UIPasteboard *pasteboard = [UIPasteboard generalPasteboard];
    pasteboard.string = self.textView.text;
    [EToast showStatus:@"Copied"];
}

- (void)imagePickerController:(UIImagePickerController *)picker didFinishPickingMediaWithInfo:(NSDictionary<NSString *,id> *)info {
    [picker dismissViewControllerAnimated:YES completion:nil];
    
    UIImage *image = (UIImage *)info[UIImagePickerControllerOriginalImage];
    
    if (!image) {
        [self.navigationController popViewControllerAnimated:YES];
    }
    
    [EToast showStatus:@"Encoding, please wait."];
    
    typeof(self) __weak weakSelf = self;
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        NSString *string = [UIImagePNGRepresentation(image) base64EncodedStringWithOptions:NSDataBase64Encoding64CharacterLineLength];
        dispatch_sync(dispatch_get_main_queue(), ^{
            __strong typeof(self) strongSelf = weakSelf;
            
            strongSelf.textView.text = string;
        });
    });
}

- (void)imagePickerControllerDidCancel:(UIImagePickerController *)picker {
    [EToast showErrorWithStatus:@"Canceled"];
    [picker dismissViewControllerAnimated:YES completion:nil];
    [self.navigationController popViewControllerAnimated:YES];
}

- (UITextView *)textView {
    if (!_textView) {
        _textView = [[UITextView alloc] init];
        [self.view addSubview:_textView];
        _textView.editable = NO;
        [_textView mas_makeConstraints:^(MASConstraintMaker *make) {
            make.edges.equalTo(self.view);
        }];
    }
    
    return _textView;
}

@end

File: encodify/Classes/Utilities/Controllers/ImageEncodeViewController.swift
//
//  ImageEncodeViewController.swift
//  encodify
//
//  Created by 洪朔 on 2024/12/20.
//  Copyright © 2024年 Tuccuay. All rights reserved.
//

import UIKit
import SnapKit

class ImageEncodeViewController: ThemeAwareViewController {
    
    private lazy var textView: UITextView = {
        let textView = UITextView()
        let baseFont = UIFont.preferredFont(forTextStyle: .body)
        textView.font = UIFont.monospacedSystemFont(ofSize: baseFont.pointSize, weight: .regular)
        textView.adjustsFontForContentSizeCategory = true
        textView.isEditable = false
        textView.backgroundColor = UIColor.encodifyCardBackground
        textView.textColor = UIColor.encodifyPrimaryText
        textView.layer.cornerRadius = 16
        textView.layer.masksToBounds = false
        textView.contentInset = UIEdgeInsets(top: 16, left: 16, bottom: 16, right: 16)
        textView.isSelectable = true
        textView.dataDetectorTypes = []
        
        // Modern shadow with enhanced depth
        textView.applyThemeAwareShadow(radius: 10, opacity: 0.12, offset: CGSize(width: 0, height: 3))
        
        // Accessibility improvements
        textView.accessibilityLabel = "Encoded image data"
        textView.accessibilityHint = "Base64 encoded image will appear here"
        
        return textView
    }()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        setupUI()
        presentImagePicker()
    }
    
    private func setupUI() {
        title = "Image Encode"
        view.backgroundColor = UIColor.systemBackground
        
        view.addSubview(textView)
        textView.snp.makeConstraints { make in
            make.edges.equalTo(view.safeAreaLayoutGuide).inset(20)
        }
        
        // Create modern copy button
        let copyButton = UIBarButtonItem(
            title: "Copy",
            style: .plain,
            target: self,
            action: #selector(copyResult)
        )
        copyButton.tintColor = UIColor.encodifyTintColor
        navigationItem.rightBarButtonItem = copyButton
        
        // Add loading indicator placeholder
        setupLoadingState()
    }
    
    private func setupLoadingState() {
        textView.text = "Loading..."
        textView.textColor = UIColor.encodifySecondaryText
    }
    
    private func presentImagePicker() {
        let imagePickerController = UIImagePickerController()
        imagePickerController.sourceType = .photoLibrary
        imagePickerController.delegate = self
        present(imagePickerController, animated: true)
    }
    
    @objc private func copyResult() {
        guard let text = textView.text, !text.isEmpty, text != "Loading..." else {
            Toast.showError("No content to copy")
            return
        }
        
        // Add haptic feedback
        let impact = UIImpactFeedbackGenerator(style: .light)
        impact.impactOccurred()
        
        UIPasteboard.general.string = text
        Toast.showStatus("Copied")
    }
}

// MARK: - UIImagePickerControllerDelegate & UINavigationControllerDelegate
extension ImageEncodeViewController: UIImagePickerControllerDelegate, UINavigationControllerDelegate {
    
    func imagePickerController(_ picker: UIImagePickerController, didFinishPickingMediaWithInfo info: [UIImagePickerController.InfoKey : Any]) {
        picker.dismiss(animated: true)
        
        guard let image = info[.originalImage] as? UIImage else {
            navigationController?.popViewController(animated: true)
            return
        }
        
        Toast.showStatus("Encoding, please wait.")
        
        DispatchQueue.global(qos: .default).async { [weak self] in
            guard let imageData = image.pngData() else { return }
            let base64String = imageData.base64EncodedString(options: .lineLength64Characters)
            
            DispatchQueue.main.async {
                self?.textView.text = base64String
                self?.textView.textColor = UIColor.encodifyPrimaryText
                
                // Add completion animation
                UIView.animate(withDuration: 0.3, delay: 0, usingSpringWithDamping: 0.8, initialSpringVelocity: 0.5, options: .allowUserInteraction) {
                    self?.textView.transform = CGAffineTransform(scaleX: 1.02, y: 1.02)
                } completion: { _ in
                    UIView.animate(withDuration: 0.2) {
                        self?.textView.transform = CGAffineTransform.identity
                    }
                }
            }
        }
    }
    
    func imagePickerControllerDidCancel(_ picker: UIImagePickerController) {
        Toast.showError("Canceled")
        picker.dismiss(animated: true)
        navigationController?.popViewController(animated: true)
    }
}
